datafusion/core/src/datasource/listing/table.rs

```diff
@@ -804,21 +804,25 @@ impl TableProvider for ListingTable {
         .await?;
 
         let file_groups = file_list_stream.try_collect::<Vec<_>>().await?;
-
-        if file_groups.len() > 1 {
-            return Err(DataFusionError::Plan(
-                "Datafusion currently supports tables from single partition and/or file."
-                    .to_owned(),
-            ));
+        let writer_mode;
+        //if we are writing a single output_partition to a table backed by a single file
+        //we can append to that file. Otherwise, we can write new files into the directory
+        //adding new files to the listing table in order to insert to the table.
+        let input_partitions = input.output_partitioning().partition_count();
+        if file_groups.len() == 1 && input_partitions == 1 {
```

Comment:
This logic works but feels inflexible to me. There may be a more explicit way for a user to express their intention, similar to Spark's [Save Modes](https://spark.apache.org/docs/latest/sql-data-sources-load-save-functions.html#save-modes) and `PartitionBy` methods.

Reply:
Are you envisioning`ListingTableWriteOptions `as part of the `ListingOptions` struct (i.e. a property of the registered table itself)? So the user would do something like:

```rust
ctx.register_csv("csv_table_sink", out_path, CsvReadOptions::default(), CsvWriteOptions::default().write_mode(WriteMode::Append))
        .await?;
```

Subsequent writes to this table would then respect the declared settings on that table. PartitionBy information is actually already included in [CsvReadOptions](https://github.com/apache/arrow-datafusion/blob/27a3c81493210435f23b64e8fc65c988bf665579/datafusion/core/src/datasource/file_format/options.rs#L71). 

I had been thinking about how to pass this option at write time:

```rust
df.write_table("csv_table_sink", WriteMode::Append)
```

but I was not sure how to propagate that additional argument from the `LogicalPlan` through to `InsertExec` (or if that even makes sense to do).

Reply:
> Are you envisioningListingTableWriteOptions as part of the ListingOptions struct (i.e. a property of the registered table itself)? So the user would do something like:

I guess I was thinking there could be a way to register the initil table with WriteOptions for when it was written to via `INSERT ...` type queries.

However, I think the more interesting usecase i my mind is passing the options at write time, as you show 

```rust
df.write_table("table", WriteOptions::new()...)
```

I am not quite sure how the code would look and how to make a nice API that separates the two concerns (the actual act of writing / passing the write options) and the registered table.

It seems like there should also be a way to write directly to a target output without having to register a table provider. I am sure there is a way we just need to come up with a clever API to dos

Reply:
> It seems like there should also be a way to write directly to a target output without having to register a table provider. I am sure there is a way we just need to come up with a clever API to do so

I definitely agree. I am thinking for that we could use the `COPY TO` SQL syntax to dump a table to files.  We will need to implement a logical plan for copy to, but it looks like that is already on your radar in #5654. The execution plan for `CopyTo` is ultimately the same as `InsertExec` which is currently created by each `FileSink`. Probably want to rename `InsertExec` to something more generic like `WriteFileSinkExec` at that point.

When that is done we can add options to `DataFrame.write_table` or create additional methods that allow hooking into a `LogicalPlanBuilder::copy_to`  method to enable writing to files without the need for a registered table.

Reply:
> . The execution plan for CopyTo is ultimately the same as InsertExec which is currently created by each FileSink. Probably want to rename InsertExec to something more generic like WriteFileSinkExec at that point.

Yes that sounds ideal